lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -785,12 +983,29 @@ class OpDependencyNode : public DependencyNode {
   }
 };
 
+/// A DependencyGraph is a DAG consisting of DependencyNodes somehow related by
+/// interaction. It contains useful metadata and functions for reasoning about
+/// and manipulating the DAG.
 class DependencyGraph {
 private:
+  // The set of root nodes in the DAG
+  // TODO: why is this a Set and `leafs` a Map? I don't really remember
```

Comment:
I suspect that they are `SetVector`s to guarantee repeatable iterating order based on insertion order and not pointer values?

Reply:
It's possible, although since we're not worrying about measurement/result ordering I think predictable iterating order has gone out the window. My suspicion is that it is a vestigial trait, and that a map will make finding roots a bit faster. Right now, I think the roots are only really iterated through when trying to find a specific root anyway.

Reply:
Having repeatable outputs with repeatable inputs is very important in my opinion, so I'd still recommend keeping the predictable iterating order.

Reply:
I'll see where they are iterated through. Scheduling sorts them currently, so if it's just when looking them up I don't see a difference. If they're iterated through elsewhere too then it would make sense.

Reply:
Ah, it's also used for code generation, so I'll leave it as a SetVector and remove the TODO.